applications/asset_tracker/src/main.c

```diff
@@ -186,14 +207,105 @@ static void shutdown_modem(void)
 #endif
 }
 
+#if defined(CONFIG_LWM2M_CARRIER)
+int lwm2m_carrier_event_handler(const lwm2m_carrier_event_t *event)
+{
+	switch (event->type) {
+	case LWM2M_CARRIER_EVENT_BSDLIB_INIT:
+		LOG_DBG("LWM2M_CARRIER_EVENT_BSDLIB_INIT");
+		k_sem_give(&bsdlib_initialized);
+		break;
+	case LWM2M_CARRIER_EVENT_CONNECTING:
+		LOG_DBG("LWM2M_CARRIER_EVENT_CONNECTING\n");
+		break;
+	case LWM2M_CARRIER_EVENT_CONNECTED:
+		LOG_DBG("LWM2M_CARRIER_EVENT_CONNECTED");
+		k_sem_give(&lte_connected);
+		break;
+	case LWM2M_CARRIER_EVENT_DISCONNECTING:
+		LOG_DBG("LWM2M_CARRIER_EVENT_DISCONNECTING");
+		break;
+	case LWM2M_CARRIER_EVENT_BOOTSTRAPPED:
+		LOG_DBG("LWM2M_CARRIER_EVENT_BOOTSTRAPPED");
+		break;
+	case LWM2M_CARRIER_EVENT_DISCONNECTED:
+		LOG_DBG("LWM2M_CARRIER_EVENT_DISCONNECTED");
+		break;
+	case LWM2M_CARRIER_EVENT_READY:
+		LOG_DBG("LWM2M_CARRIER_EVENT_READY");
+		k_sem_give(&cloud_ready_to_connect);
+		break;
+	case LWM2M_CARRIER_EVENT_FOTA_START:
+		LOG_DBG("LWM2M_CARRIER_EVENT_FOTA_START");
+		/* Due to limitations in the number of secure sockets, the cloud
+		 * socket has to be closed when the carrier library initiates
+		 * firmware upgrade download.
+		 */
+		atomic_set(&carrier_requested_disconnect, 1);
+		app_disconnect();
+		break;
+	case LWM2M_CARRIER_EVENT_REBOOT:
+		LOG_DBG("LWM2M_CARRIER_EVENT_REBOOT");
+		break;
+	case LWM2M_CARRIER_EVENT_ERROR:
+		LOG_ERR("LWM2M_CARRIER_EVENT_ERROR: code %d, value %d",
+			((lwm2m_carrier_event_error_t *)event->data)->code,
+			((lwm2m_carrier_event_error_t *)event->data)->value);
+		break;
+	}
+
+	return 0;
+}
+
+/**@brief Disconnects from cloud. First it tries using the cloud backend's
+ *	  disconnect() implementation. If that fails, it falls back to close the
+ *	  socket directly, using close().
+ */
+static void app_disconnect(void)
+{
+	int err;
+
+	atomic_set(&cloud_association, CLOUD_ASSOCIATION_STATE_INIT);
+	LOG_INF("Disconnecting from cloud.");
+
+	err = cloud_disconnect(cloud_backend);
+	if (err == -ENOTCONN) {
+		LOG_INF("Cloud connection was not established.");
+		return;
+	}
+
+	if (err) {
+		LOG_ERR("Could not disconnect from cloud, err: %d", err);
+		LOG_INF("Closing the cloud socket directly.");
+
+		err = close(cloud_backend->config->socket);
+		if (err) {
+			LOG_ERR("Failed to close socket, error: %d", err);
+			return;
+		}
+
+		LOG_INF("Socket was closed successfully.");
+		return;
+	}
+
+	/* Ensure that the socket is indeed closed before returning. */
+	k_sem_take(&cloud_disconnected, K_FOREVER);
```

Comment:
Is there a possibility we would hang here forever? Can't see anything obvious, but it might have overlooked something

Reply:
forever is a long time... preference on a timeout value?  30s?
this is currently called from the carrier lib callback, so it's blocking the lib.